// @prisma/migrate/dist/commands/MigrateDev.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  MigrateDev: () => MigrateDev
});
var import_sdk = __toModule(require("@prisma/sdk"));
var import_debug = __toModule(require("@prisma/debug"));
var import_chalk = __toModule(require("chalk"));
var import_prompts = __toModule(require("prompts"));
var import_fs = __toModule(require("fs"));
var import_path = __toModule(require("path"));
var import_Migrate = __toModule(require("../Migrate"));
var import_ensureDatabaseExists = __toModule(require("../utils/ensureDatabaseExists"));
var import_flagErrors = __toModule(require("../utils/flagErrors"));
var import_errors = __toModule(require("../utils/errors"));
var import_printMigrationId = __toModule(require("../utils/printMigrationId"));
var import_printFiles = __toModule(require("../utils/printFiles"));
var import_handleEvaluateDataloss = __toModule(require("../utils/handleEvaluateDataloss"));
var import_promptForMigrationName = __toModule(require("../utils/promptForMigrationName"));
var import_detectOldMigrate = __toModule(require("../utils/detectOldMigrate"));
var import_printDatasource = __toModule(require("../utils/printDatasource"));
var import_seed = __toModule(require("../utils/seed"));
const debug = (0, import_debug.default)("prisma:migrate:dev");
const _MigrateDev = class {
  static new() {
    return new _MigrateDev();
  }
  async parse(argv) {
    var _a, _b;
    const args = (0, import_sdk.arg)(argv, {
      "--help": Boolean,
      "-h": "--help",
      "--name": String,
      "-n": "--name",
      "--create-only": Boolean,
      "--schema": String,
      "--skip-generate": Boolean,
      "--skip-seed": Boolean,
      "--experimental": Boolean,
      "--early-access-feature": Boolean,
      "--telemetry-information": String
    });
    if ((0, import_sdk.isError)(args)) {
      return this.help(args.message);
    }
    if (args["--help"]) {
      return this.help();
    }
    if (args["--experimental"]) {
      throw new import_flagErrors.ExperimentalFlagWithNewMigrateError();
    }
    if (args["--early-access-feature"]) {
      throw new import_flagErrors.EarlyAccessFeatureFlagWithNewMigrateError();
    }
    const schemaPath = await (0, import_sdk.getSchemaPath)(args["--schema"]);
    if (!schemaPath) {
      throw new import_errors.NoSchemaFoundError();
    }
    console.info(import_chalk.default.dim(`Prisma schema loaded from ${import_path.default.relative(process.cwd(), schemaPath)}`));
    await (0, import_printDatasource.printDatasource)(schemaPath);
    console.info();
    (0, import_detectOldMigrate.throwUpgradeErrorIfOldMigrate)(schemaPath);
    const schema = import_fs.default.readFileSync(schemaPath, "utf-8");
    await (0, import_sdk.getDMMF)({
      datamodel: schema
    });
    await (0, import_sdk.getConfig)({
      datamodel: schema
    });
    const wasDbCreated = await (0, import_ensureDatabaseExists.ensureDatabaseExists)("create", true, schemaPath);
    if (wasDbCreated) {
      console.info(wasDbCreated);
    }
    const migrate = new import_Migrate.Migrate(schemaPath);
    let devDiagnostic;
    try {
      devDiagnostic = await migrate.devDiagnostic();
      debug({ devDiagnostic: JSON.stringify(devDiagnostic, null, 2) });
    } catch (e) {
      migrate.stop();
      throw e;
    }
    const migrationIdsApplied = [];
    if (devDiagnostic.action.tag === "reset") {
      if (!args["--force"]) {
        if ((0, import_sdk.isCi)() && Boolean((_a = import_prompts.default._injected) == null ? void 0 : _a.length) === false) {
          migrate.stop();
          throw new import_errors.MigrateDevEnvNonInteractiveError();
        }
        const dbInfo = await (0, import_ensureDatabaseExists.getDbInfo)(schemaPath);
        const confirmedReset = await this.confirmReset(dbInfo, devDiagnostic.action.reason);
        console.info();
        if (!confirmedReset) {
          console.info("Reset cancelled.");
          migrate.stop();
          process.exit(0);
          return ``;
        }
      }
      try {
        await migrate.reset();
      } catch (e) {
        migrate.stop();
        throw e;
      }
    }
    try {
      const { appliedMigrationNames } = await migrate.applyMigrations();
      migrationIdsApplied.push(...appliedMigrationNames);
      if (appliedMigrationNames.length > 0) {
        console.info(`The following migration(s) have been applied:

${(0, import_chalk.default)((0, import_printFiles.printFilesFromMigrationIds)("migrations", appliedMigrationNames, {
          "migration.sql": ""
        }))}`);
      }
    } catch (e) {
      migrate.stop();
      throw e;
    }
    if (devDiagnostic.action.tag === "reset" && !process.env.PRISMA_MIGRATE_SKIP_SEED && !args["--skip-seed"]) {
      try {
        const detected = (0, import_seed.detectSeedFiles)(schemaPath);
        if (detected.numberOfSeedFiles > 0) {
          console.info();
          await (0, import_seed.tryToRunSeed)(schemaPath);
        }
      } catch (e) {
        console.error(e);
      }
    }
    let evaluateDataLossResult;
    try {
      evaluateDataLossResult = await migrate.evaluateDataLoss();
      debug({ evaluateDataLossResult });
    } catch (e) {
      migrate.stop();
      throw e;
    }
    const unexecutableStepsError = (0, import_handleEvaluateDataloss.handleUnexecutableSteps)(evaluateDataLossResult.unexecutableSteps, args["--create-only"]);
    if (unexecutableStepsError) {
      migrate.stop();
      throw new Error(unexecutableStepsError);
    }
    if (evaluateDataLossResult.warnings && evaluateDataLossResult.warnings.length > 0) {
      console.log(import_chalk.default.bold(`
\u26A0\uFE0F  Warnings for the current datasource:
`));
      for (const warning of evaluateDataLossResult.warnings) {
        console.log((0, import_chalk.default)(`  \u2022 ${warning.message}`));
      }
      console.info();
      if (!args["--force"]) {
        if ((0, import_sdk.isCi)() && Boolean((_b = import_prompts.default._injected) == null ? void 0 : _b.length) === false) {
          migrate.stop();
          throw new import_errors.MigrateDevEnvNonInteractiveError();
        }
        const message = args["--create-only"] ? "Are you sure you want create this migration?" : "Are you sure you want create and apply this migration?";
        const confirmation = await (0, import_prompts.default)({
          type: "confirm",
          name: "value",
          message
        });
        if (!confirmation.value) {
          migrate.stop();
          return `Migration cancelled.`;
        }
      }
    }
    let migrationName = void 0;
    if (evaluateDataLossResult.migrationSteps > 0 || args["--create-only"]) {
      const getMigrationNameResult = await (0, import_promptForMigrationName.getMigrationName)(args["--name"]);
      if (getMigrationNameResult.userCancelled) {
        migrate.stop();
        return getMigrationNameResult.userCancelled;
      } else {
        migrationName = getMigrationNameResult.name;
      }
    }
    let migrationIds;
    try {
      const createMigrationResult = await migrate.createMigration({
        migrationsDirectoryPath: migrate.migrationsDirectoryPath,
        migrationName: migrationName || "",
        draft: args["--create-only"] ? true : false,
        prismaSchema: migrate.getDatamodel()
      });
      debug({ createMigrationResult });
      if (args["--create-only"]) {
        migrate.stop();
        return `Prisma Migrate created the following migration without applying it ${(0, import_printMigrationId.printMigrationId)(createMigrationResult.generatedMigrationName)}

You can now edit it and apply it by running ${import_chalk.default.greenBright((0, import_sdk.getCommandWithExecutor)("prisma migrate dev"))}.`;
      }
      const { appliedMigrationNames } = await migrate.applyMigrations();
      migrationIds = appliedMigrationNames;
    } finally {
      migrate.stop();
    }
    migrationIdsApplied.length > 0 && console.info();
    if (migrationIds.length === 0) {
      if (migrationIdsApplied.length > 0) {
        console.info(`${import_chalk.default.green("Your database is now in sync with your schema.")}`);
      } else {
        console.info(`Already in sync, no schema change or pending migration was found.`);
      }
    } else {
      console.info(`The following migration(s) have been created and applied from new schema changes:

${(0, import_chalk.default)((0, import_printFiles.printFilesFromMigrationIds)("migrations", migrationIds, {
        "migration.sql": ""
      }))}

${import_chalk.default.green("Your database is now in sync with your schema.")}`);
    }
    if (!process.env.PRISMA_MIGRATE_SKIP_GENERATE && !args["--skip-generate"]) {
      await migrate.tryToRunGenerate();
      console.info();
    }
    return "";
  }
  async confirmReset({
    schemaWord,
    dbType,
    dbName,
    dbLocation
  }, reason) {
    const mssqlMessage = `${reason}

We need to reset the database.
Do you want to continue? ${import_chalk.default.red("All data will be lost")}.`;
    const message = `${reason}

We need to reset the ${dbType} ${schemaWord} "${dbName}" at "${dbLocation}".
Do you want to continue? ${import_chalk.default.red("All data will be lost")}.`;
    const confirmation = await (0, import_prompts.default)({
      type: "confirm",
      name: "value",
      message: dbType === "SQL Server" ? mssqlMessage : message
    });
    return confirmation.value;
  }
  help(error) {
    if (error) {
      return new import_sdk.HelpError(`
${import_chalk.default.bold.red(`!`)} ${error}
${_MigrateDev.help}`);
    }
    return _MigrateDev.help;
  }
};
let MigrateDev = _MigrateDev;
MigrateDev.help = (0, import_sdk.format)(`
${process.platform === "win32" ? "" : import_chalk.default.bold("\u{1F3CB}\uFE0F  ")}Create a migration from changes in Prisma schema, apply it to the database, trigger generators (e.g. Prisma Client)
 
${import_chalk.default.bold("Usage")}

  ${import_chalk.default.dim("$")} prisma migrate dev [options]

${import_chalk.default.bold("Options")}

       -h, --help   Display this help message
         --schema   Custom path to your Prisma schema
       -n, --name   Name the migration
    --create-only   Create a new migration but do not apply it
                    The migration will be empty if there are no changes in Prisma schema
  --skip-generate   Skip triggering generators (e.g. Prisma Client)
      --skip-seed   Skip triggering seed

${import_chalk.default.bold("Examples")}

  Create a migration from changes in Prisma schema, apply it to the database, trigger generators (e.g. Prisma Client)
  ${import_chalk.default.dim("$")} prisma migrate dev

  Specify a schema
  ${import_chalk.default.dim("$")} prisma migrate dev --schema=./schema.prisma

  Create a migration without applying it
  ${import_chalk.default.dim("$")} prisma migrate dev --create-only
  `);
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  MigrateDev
});
