// @keystone-next/keystone/testing/dist/keystone.esm.js
import _objectSpread from '@babel/runtime/helpers/esm/objectSpread2';
import Path__default from 'path';
import crypto from 'crypto';
import fs from 'fs';
import supertest from 'supertest';
import memoizeOne from 'memoize-one';
import { getCommittedArtifacts, writeCommittedArtifacts, generateNodeModulesArtifacts, requirePrismaClient } from '../../artifacts/dist/keystone.esm.js';
import { p as pushPrismaSchemaToDatabase } from '../../dist/migrations-7717e45c.esm.js';
import { c as createSystem } from '../../dist/createSystem-0871d001.esm.js';
import { c as createExpressServer } from '../../dist/createAdminUIMiddleware-456cab36.esm.js';
import { i as initConfig } from '../../dist/initConfig-f536b388.esm.js';
import 'apollo-server-micro';
import 'apollo-server-express';
import '@babel/runtime/helpers/objectWithoutProperties';
import 'graphql';
import 'cookie';
import '@hapi/iron';
import 'uid-safe';
import '../../dist/sqlite-943cbaa7.esm.js';
import 'fs-extra';
import '@prisma/sdk';
import 'prettier';
import '../../dist/utils-b84e12de.esm.js';
import 'prompts';
import '@babel/runtime/helpers/defineProperty';
import '../../dist/core-3f0f7b15.esm.js';
import 'decimal.js';
import '@graphql-ts/schema/api-without-context';
import '@graphql-ts/schema';
import 'graphql-type-json';
import 'graphql-upload/public/GraphQLUpload.js';
import '@graphql-ts/schema/api-with-context';
import 'fast-glob';
import 'resolve';
import '@emotion/hash';
import '../../dist/admin-meta-graphql-c4fe434c.esm.js';
import '@apollo/client';
import '../../session/dist/keystone.esm.js';
import '@babel/runtime/helpers/classPrivateFieldInitSpec';
import '@babel/runtime/helpers/classPrivateFieldGet';
import '@babel/runtime/helpers/classPrivateFieldSet';
import 'p-limit';
import 'pluralize';
import 'apollo-server-errors';
import 'uuid';
import 'image-type';
import 'image-size';
import '../../fields/types/image/utils/dist/keystone.esm.js';
import 'node-fetch';
import 'form-data';
import 'stream';
import 'filenamify';
import '@sindresorhus/slugify';
import '../../fields/types/file/utils/dist/keystone.esm.js';
import '@prisma/migrate';
import 'chalk';
import 'cors';
import 'express';
import 'graphql-upload';
import '../../dist/createApolloServer-7b3e0bd8.esm.js';
import 'url';
import 'cuid';
import '../../dist/package-path-7ba957ae.esm.js';

const _hashPrismaSchema = memoizeOne(prismaSchema => crypto.createHash('md5').update(prismaSchema).digest('hex'));

const _alreadyGeneratedProjects = new Set();

async function setupTestEnv({
  config: _config
}) {
  // Force the UI to always be disabled.
  const config = initConfig(_objectSpread(_objectSpread({}, _config), {}, {
    ui: {
      isDisabled: true
    }
  }));
  const {
    graphQLSchema,
    getKeystone
  } = createSystem(config);
  const artifacts = await getCommittedArtifacts(graphQLSchema, config);

  const hash = _hashPrismaSchema(artifacts.prisma);

  const artifactPath = Path__default.resolve('.keystone', 'tests', hash);

  if (!_alreadyGeneratedProjects.has(hash)) {
    _alreadyGeneratedProjects.add(hash);

    fs.mkdirSync(artifactPath, {
      recursive: true
    });
    await writeCommittedArtifacts(artifacts, artifactPath);
    await generateNodeModulesArtifacts(graphQLSchema, config, artifactPath);
  }

  await pushPrismaSchemaToDatabase(config.db.url, artifacts.prisma, Path__default.join(artifactPath, 'schema.prisma'), true // shouldDropDatabase
  );
  const {
    connect,
    disconnect,
    createContext
  } = getKeystone(requirePrismaClient(artifactPath)); // (config, graphQLSchema, createContext, dev, projectAdminPath, isVerbose)

  const app = await createExpressServer(config, graphQLSchema, createContext);

  const graphQLRequest = ({
    query,
    variables = undefined,
    operationName
  }) => {
    var _config$graphql;

    return supertest(app).post(((_config$graphql = config.graphql) === null || _config$graphql === void 0 ? void 0 : _config$graphql.path) || '/api/graphql').send({
      query,
      variables,
      operationName
    }).set('Accept', 'application/json');
  };

  return {
    connect,
    disconnect,
    testArgs: {
      context: createContext(),
      graphQLRequest,
      app
    }
  };
}
function setupTestRunner({
  config
}) {
  return testFn => async () => {
    // Reset the database to be empty for every test.
    const {
      connect,
      disconnect,
      testArgs
    } = await setupTestEnv({
      config
    });
    await connect();

    try {
      return await testFn(testArgs);
    } finally {
      await disconnect();
    }
  };
}

export { setupTestEnv, setupTestRunner };
