// @keystone-next/keystone/testing/dist/keystone.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var Path = require('path');
var crypto = require('crypto');
var fs = require('fs');
var supertest = require('supertest');
var memoizeOne = require('memoize-one');
var artifacts_dist_keystone = require('../../artifacts/dist/keystone.cjs.dev.js');
var migrations = require('../../dist/migrations-e3f537e1.cjs.dev.js');
var createSystem = require('../../dist/createSystem-558f626f.cjs.dev.js');
var createAdminUIMiddleware = require('../../dist/createAdminUIMiddleware-7ca43618.cjs.dev.js');
var initConfig = require('../../dist/initConfig-d774ef5b.cjs.dev.js');
require('apollo-server-micro');
require('apollo-server-express');
require('@babel/runtime/helpers/objectWithoutProperties');
require('graphql');
require('cookie');
require('@hapi/iron');
require('uid-safe');
require('../../dist/sqlite-b63545bc.cjs.dev.js');
require('fs-extra');
require('@prisma/sdk');
require('prettier');
require('../../dist/utils-de2e2c01.cjs.dev.js');
require('prompts');
require('@babel/runtime/helpers/defineProperty');
require('../../dist/core-c8ecce23.cjs.dev.js');
require('decimal.js');
require('@graphql-ts/schema/api-without-context');
require('@graphql-ts/schema');
require('graphql-type-json');
require('graphql-upload/public/GraphQLUpload.js');
require('@graphql-ts/schema/api-with-context');
require('fast-glob');
require('resolve');
require('@emotion/hash');
require('../../dist/admin-meta-graphql-1e30b380.cjs.dev.js');
require('@apollo/client');
require('../../session/dist/keystone.cjs.dev.js');
require('@babel/runtime/helpers/classPrivateFieldInitSpec');
require('@babel/runtime/helpers/classPrivateFieldGet');
require('@babel/runtime/helpers/classPrivateFieldSet');
require('p-limit');
require('pluralize');
require('apollo-server-errors');
require('uuid');
require('image-type');
require('image-size');
require('../../fields/types/image/utils/dist/keystone.cjs.dev.js');
require('node-fetch');
require('form-data');
require('stream');
require('filenamify');
require('@sindresorhus/slugify');
require('../../fields/types/file/utils/dist/keystone.cjs.dev.js');
require('@prisma/migrate');
require('chalk');
require('cors');
require('express');
require('graphql-upload');
require('../../dist/createApolloServer-8725cc81.cjs.dev.js');
require('url');
require('cuid');
require('../../dist/package-path-d511a573.cjs.dev.js');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var Path__default = /*#__PURE__*/_interopDefault(Path);
var crypto__default = /*#__PURE__*/_interopDefault(crypto);
var fs__default = /*#__PURE__*/_interopDefault(fs);
var supertest__default = /*#__PURE__*/_interopDefault(supertest);
var memoizeOne__default = /*#__PURE__*/_interopDefault(memoizeOne);

const _hashPrismaSchema = memoizeOne__default['default'](prismaSchema => crypto__default['default'].createHash('md5').update(prismaSchema).digest('hex'));

const _alreadyGeneratedProjects = new Set();

async function setupTestEnv({
  config: _config
}) {
  // Force the UI to always be disabled.
  const config = initConfig.initConfig(_objectSpread(_objectSpread({}, _config), {}, {
    ui: {
      isDisabled: true
    }
  }));
  const {
    graphQLSchema,
    getKeystone
  } = createSystem.createSystem(config);
  const artifacts = await artifacts_dist_keystone.getCommittedArtifacts(graphQLSchema, config);

  const hash = _hashPrismaSchema(artifacts.prisma);

  const artifactPath = Path__default['default'].resolve('.keystone', 'tests', hash);

  if (!_alreadyGeneratedProjects.has(hash)) {
    _alreadyGeneratedProjects.add(hash);

    fs__default['default'].mkdirSync(artifactPath, {
      recursive: true
    });
    await artifacts_dist_keystone.writeCommittedArtifacts(artifacts, artifactPath);
    await artifacts_dist_keystone.generateNodeModulesArtifacts(graphQLSchema, config, artifactPath);
  }

  await migrations.pushPrismaSchemaToDatabase(config.db.url, artifacts.prisma, Path__default['default'].join(artifactPath, 'schema.prisma'), true // shouldDropDatabase
  );
  const {
    connect,
    disconnect,
    createContext
  } = getKeystone(artifacts_dist_keystone.requirePrismaClient(artifactPath)); // (config, graphQLSchema, createContext, dev, projectAdminPath, isVerbose)

  const app = await createAdminUIMiddleware.createExpressServer(config, graphQLSchema, createContext);

  const graphQLRequest = ({
    query,
    variables = undefined,
    operationName
  }) => {
    var _config$graphql;

    return supertest__default['default'](app).post(((_config$graphql = config.graphql) === null || _config$graphql === void 0 ? void 0 : _config$graphql.path) || '/api/graphql').send({
      query,
      variables,
      operationName
    }).set('Accept', 'application/json');
  };

  return {
    connect,
    disconnect,
    testArgs: {
      context: createContext(),
      graphQLRequest,
      app
    }
  };
}
function setupTestRunner({
  config
}) {
  return testFn => async () => {
    // Reset the database to be empty for every test.
    const {
      connect,
      disconnect,
      testArgs
    } = await setupTestEnv({
      config
    });
    await connect();

    try {
      return await testFn(testArgs);
    } finally {
      await disconnect();
    }
  };
}

exports.setupTestEnv = setupTestEnv;
exports.setupTestRunner = setupTestRunner;
