// @graphql-ts/schema/dist/graphql-ts-schema.esm.js
import { f as field, o as object, a as fields, i as interfaceType, b as interfaceField, u as union } from './api-with-context-832081d4.esm.js';
export { c as bindGraphQLSchemaAPIToContext } from './api-with-context-832081d4.esm.js';
import { a as arg, B as Boolean, F as Float, I as ID, b as Int, S as String, i as inputObject, l as list, n as nonNull, e as enumType, c as enumValues, s as scalar } from './list-and-non-null-5cb960d7.esm.js';
import 'graphql/type/definition';
import 'graphql/type/scalars';

/**
 * The `graphql` export is the primary entrypoint into `@graphql-ts/schema` that
 * lets you compose GraphQL types into a GraphQL Schema
 *
 * A simple schema with only a query type looks like this.
 *
 * ```ts
 * import { graphql } from "@graphql-ts/schema";
 * import { GraphQLSchema, graphql as runGraphQL } from "graphql";
 *
 * const Query = graphql.object()({
 *   name: "Query",
 *   fields: {
 *     hello: graphql.field({
 *       type: graphql.String,
 *       resolve() {
 *         return "Hello!";
 *       },
 *     }),
 *   },
 * });
 *
 * const schema = new GraphQLSchema({
 *   query: Query.graphQLType,
 * });
 *
 * runGraphQL({
 *   source: `
 *     query {
 *       hello
 *     }
 *   `,
 *   schema,
 * }).then((result) => {
 *   console.log(result);
 * });
 * ```
 *
 * You can use pass the `graphQLSchema` to `ApolloServer` and other GraphQL servers.
 *
 * You can also create a more advanced schema with other object types, args and mutations.
 *
 * ```ts
 * import { graphql } from "@graphql-ts/schema";
 * import { GraphQLSchema, graphql as runGraphQL } from "graphql";
 * import { deepEqual } from "assert";
 *
 * type TodoItem = {
 *   title: string;
 * };
 *
 * const todos: TodoItem[] = [];
 *
 * const Todo = graphql.object<TodoItem>({
 *   name: "Todo",
 *   fields: {
 *     title: graphql.field({ type: graphql.String }),
 *   },
 * });
 *
 * const Query = graphql.object()({
 *   name: "Query",
 *   fields: {
 *     todos: graphql.field({
 *       type: graphql.list(Todo),
 *       resolve() {
 *         return todos;
 *       },
 *     }),
 *   },
 * });
 *
 * const Mutation = graphql.object()({
 *   name: "Mutation",
 *   fields: {
 *     createTodo: graphql.field({
 *       args: {
 *         title: graphql.arg({ type: graphql.String }),
 *       },
 *       type: Todo,
 *       resolve(rootVal, { title }) {
 *         const todo = { title };
 *         todos.push(todo);
 *         return todo;
 *       },
 *     }),
 *   },
 * });
 *
 * const schema = new GraphQLSchema({
 *   query: Query.graphQLType,
 *   mutation: Mutation.graphQLType,
 * });
 *
 * (async () => {
 *   const result = await runGraphQL({
 *     source: `
 *       query {
 *         todos {
 *           title
 *         }
 *       }
 *     `,
 *     schema,
 *   });
 *   deepEqual(result, { data: { todos: [] } });
 *
 *   const result = await runGraphQL({
 *     source: `
 *       mutation {
 *         createTodo(title: "Try @graphql-ts/schema") {
 *           title
 *         }
 *       }
 *     `,
 *     schema,
 *   });
 *   deepEqual(result, {
 *     data: { createTodo: { title: "Try @graphql-ts/schema" } },
 *   });
 *
 *   const result = await runGraphQL({
 *     source: `
 *       query {
 *         todos {
 *           title
 *         }
 *       }
 *     `,
 *     schema,
 *   });
 *   deepEqual(result, {
 *     data: { todos: [{ title: "Try @graphql-ts/schema" }] },
 *   });
 * })();
 * ```
 *
 * For information on how to construct other types like input objects, unions,
 * interfaces and enums and more detailed information, see the documentation in
 * the `graphql` export below.
 *
 * When using it, you're going to want to create your own version of it bound to
 * your specific `Context` type.
 *
 * @module
 */
/**
 * The particular `Context` type for this `graphql` export that is provided to
 * field resolvers.
 *
 * Below, there are many types exported which are similar to the types with the
 * same names exported from the root of the package except that they are bound
 * to the `Context` type so they can be used elsewhere without needing to be
 * bound to the context on every usage.
 */

var schemaApi = /*#__PURE__*/Object.freeze({
  __proto__: null,
  field: field,
  object: object,
  arg: arg,
  Boolean: Boolean,
  Float: Float,
  ID: ID,
  Int: Int,
  String: String,
  inputObject: inputObject,
  list: list,
  nonNull: nonNull,
  'enum': enumType,
  enumValues: enumValues,
  fields: fields,
  'interface': interfaceType,
  interfaceField: interfaceField,
  union: union,
  scalar: scalar
});

export { schemaApi as graphql };
