// @prisma/migrate/dist/Migrate.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  Migrate: () => Migrate
});
var import_sdk = __toModule(require("@prisma/sdk"));
var import_sdk2 = __toModule(require("@prisma/sdk"));
var import_chalk = __toModule(require("chalk"));
var import_debug = __toModule(require("@prisma/debug"));
var import_fs = __toModule(require("fs"));
var import_log_update = __toModule(require("log-update"));
var import_path = __toModule(require("path"));
var import_MigrateEngine = __toModule(require("./MigrateEngine"));
var import_formatms = __toModule(require("./utils/formatms"));
var import_engines_version = __toModule(require("@prisma/engines-version"));
var import_errors = __toModule(require("./utils/errors"));
const debug = (0, import_debug.default)("prisma:migrate");
const packageJson = eval(`require('../package.json')`);
class Migrate {
  get devMigrationsDir() {
    return import_path.default.join(import_path.default.dirname(this.schemaPath), "migrations/dev");
  }
  constructor(schemaPath, enabledPreviewFeatures) {
    this.schemaPath = this.getSchemaPath(schemaPath);
    this.migrationsDirectoryPath = import_path.default.join(import_path.default.dirname(this.schemaPath), "migrations");
    this.engine = new import_MigrateEngine.MigrateEngine({
      projectDir: import_path.default.dirname(this.schemaPath),
      schemaPath: this.schemaPath,
      enabledPreviewFeatures
    });
  }
  stop() {
    this.engine.stop();
  }
  getSchemaPath(schemaPathFromOptions) {
    const schemaPath = (0, import_sdk2.getSchemaPathSync)(schemaPathFromOptions);
    if (!schemaPath) {
      throw new import_errors.NoSchemaFoundError();
    }
    return schemaPath;
  }
  getDatamodel() {
    return import_fs.default.readFileSync(this.schemaPath, "utf-8");
  }
  reset() {
    return this.engine.reset();
  }
  createMigration(params) {
    return this.engine.createMigration(params);
  }
  diagnoseMigrationHistory({
    optInToShadowDatabase
  }) {
    return this.engine.diagnoseMigrationHistory({
      migrationsDirectoryPath: this.migrationsDirectoryPath,
      optInToShadowDatabase
    });
  }
  listMigrationDirectories() {
    return this.engine.listMigrationDirectories({
      migrationsDirectoryPath: this.migrationsDirectoryPath
    });
  }
  devDiagnostic() {
    return this.engine.devDiagnostic({
      migrationsDirectoryPath: this.migrationsDirectoryPath
    });
  }
  async markMigrationApplied({
    migrationId
  }) {
    return await this.engine.markMigrationApplied({
      migrationsDirectoryPath: this.migrationsDirectoryPath,
      migrationName: migrationId
    });
  }
  markMigrationRolledBack({
    migrationId
  }) {
    return this.engine.markMigrationRolledBack({
      migrationName: migrationId
    });
  }
  applyMigrations() {
    return this.engine.applyMigrations({
      migrationsDirectoryPath: this.migrationsDirectoryPath
    });
  }
  evaluateDataLoss() {
    const datamodel = this.getDatamodel();
    return this.engine.evaluateDataLoss({
      migrationsDirectoryPath: this.migrationsDirectoryPath,
      prismaSchema: datamodel
    });
  }
  async push({
    force = false
  }) {
    const datamodel = this.getDatamodel();
    const { warnings, unexecutable, executedSteps } = await this.engine.schemaPush({
      force,
      schema: datamodel
    });
    return {
      executedSteps,
      warnings,
      unexecutable
    };
  }
  async tryToRunGenerate() {
    var _a;
    const message = [];
    console.info();
    (0, import_log_update.default)(`Running generate... ${import_chalk.default.dim("(Use --skip-generate to skip the generators)")}`);
    const generators = await (0, import_sdk2.getGenerators)({
      schemaPath: this.schemaPath,
      printDownloadProgress: true,
      version: import_engines_version.enginesVersion,
      cliVersion: packageJson.version
    });
    for (const generator of generators) {
      const toStr = generator.options.generator.output ? import_chalk.default.dim(` to .${import_path.default.sep}${import_path.default.relative(process.cwd(), (0, import_sdk.parseEnvValue)(generator.options.generator.output))}`) : "";
      const name = generator.manifest ? generator.manifest.prettyName : (0, import_sdk.parseEnvValue)(generator.options.generator.provider);
      (0, import_log_update.default)(`Running generate... - ${name}`);
      const before = Date.now();
      try {
        await generator.generate();
        const after = Date.now();
        const version = (_a = generator.manifest) == null ? void 0 : _a.version;
        message.push(`\u2714 Generated ${import_chalk.default.bold(name)}${version ? ` (${version})` : ""}${toStr} in ${(0, import_formatms.formatms)(after - before)}`);
        generator.stop();
      } catch (err) {
        message.push(`${err.message}`);
        generator.stop();
      }
    }
    (0, import_log_update.default)(message.join("\n"));
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Migrate
});
