// @prisma/migrate/dist/utils/handlePanic.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __markAsModule = (target) => __defProp(target, "__esModule", { value: true });
var __export = (target, all) => {
  __markAsModule(target);
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __reExport = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, { get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable });
  }
  return target;
};
var __toModule = (module2) => {
  return __reExport(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? { get: () => module2.default, enumerable: true } : { value: module2, enumerable: true })), module2);
};
__export(exports, {
  handlePanic: () => handlePanic
});
var import_chalk = __toModule(require("chalk"));
var import_sdk = __toModule(require("@prisma/sdk"));
var import_prompts = __toModule(require("prompts"));
var import_getGithubIssueUrl = __toModule(require("./getGithubIssueUrl"));
async function handlePanic(error, cliVersion, binaryVersion, command) {
  var _a;
  if ((0, import_sdk.isCi)() && Boolean((_a = import_prompts.default._injected) == null ? void 0 : _a.length) === false) {
    throw error;
  }
  await panicDialog(error, cliVersion, binaryVersion, command);
}
async function panicDialog(error, cliVersion, binaryVersion, command) {
  const errorMessage = error.message.split("\n").slice(0, Math.max(20, process.stdout.rows)).join("\n");
  console.log(`${import_chalk.default.red("Oops, an unexpected error occured!")}
${import_chalk.default.red(errorMessage)}

${import_chalk.default.bold("Please help us improve Prisma by submitting an error report.")}
${import_chalk.default.bold("Error reports never contain personal or other sensitive information.")}
${import_chalk.default.dim(`Learn more: ${(0, import_sdk.link)("https://pris.ly/d/telemetry")}`)}
`);
  const response = await (0, import_prompts.default)({
    type: "select",
    name: "value",
    message: "Submit error report",
    initial: 0,
    choices: [
      {
        title: "Yes",
        value: true,
        description: `Send error report once`
      },
      {
        title: "No",
        value: false,
        description: `Don't send error report`
      }
    ]
  });
  const reportFailedMessage = `${import_chalk.default.bold.red("Oops. We could not send the error report.")}`;
  if (response.value) {
    let reportId;
    try {
      console.log("Submitting...");
      reportId = await (0, import_sdk.sendPanic)(error, cliVersion, binaryVersion);
    } catch (error2) {
      console.log(reportFailedMessage);
    }
    if (reportId) {
      console.log(`
${import_chalk.default.bold(`We successfully received the error report id: ${reportId}`)}`);
      console.log(`
${import_chalk.default.bold("Thanks a lot for your help! \u{1F64F}")}`);
    }
  }
  await (0, import_getGithubIssueUrl.wouldYouLikeToCreateANewIssue)({
    prompt: !response.value,
    error,
    cliVersion,
    binaryVersion,
    command
  });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  handlePanic
});
