// @keystone-next/keystone/___internal-do-not-use-will-break-in-patch/admin-ui/pages/HomePage/dist/keystone.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _extends = require('@babel/runtime/helpers/extends');
var React = require('react');
var core = require('@keystone-ui/core');
var PlusIcon = require('@keystone-ui/icons/icons/PlusIcon');
var modals = require('@keystone-ui/modals');
var loading = require('@keystone-ui/loading');
var dataGetter = require('../../../../../dist/dataGetter-5d5b5868.cjs.prod.js');
require('@babel/runtime/helpers/objectSpread2');
require('../../../../../dist/getRootGraphQLFieldsFromFieldController-1358c36a.cjs.prod.js');
require('fast-deep-equal');
var CreateItemDrawer = require('../../../../../dist/CreateItemDrawer-250f084d.cjs.prod.js');
var PageContainer = require('../../../../../dist/PageContainer-4876081c.cjs.prod.js');
var client = require('@apollo/client');
var adminUi_context_dist_keystone = require('../../../../../admin-ui/context/dist/keystone.cjs.prod.js');
var adminUi_router_dist_keystone = require('../../../../../admin-ui/router/dist/keystone.cjs.prod.js');
var router = require('next/router');
require('@emotion/weak-memoize');
require('graphql');
require('@keystone-ui/toast');
require('../../../../../dist/Fields-052fa157.cjs.prod.js');
require('../../../../../dist/GraphQLErrorNotice-c553a755.cjs.prod.js');
require('@keystone-ui/notice');
require('@babel/runtime/helpers/objectWithoutProperties');
require('@keystone-ui/button');
require('@keystone-ui/popover');
require('@keystone-ui/icons/icons/MoreHorizontalIcon');
require('@keystone-ui/icons/icons/ChevronRightIcon');
require('../../../../../dist/SignoutButton-abc7d471.cjs.prod.js');
require('next/link');
require('apollo-upload-client');
require('@emotion/hash');
require('../../../../../dist/core-fe91e1fc.cjs.prod.js');
require('../../../../../dist/sqlite-bb0466dc.cjs.prod.js');
require('decimal.js');
require('@graphql-ts/schema/api-without-context');
require('@graphql-ts/schema');
require('graphql-type-json');
require('graphql-upload/public/GraphQLUpload.js');
require('@graphql-ts/schema/api-with-context');
require('../../../../../dist/admin-meta-graphql-050d49f5.cjs.prod.js');

const ListCard = ({
  listKey,
  count
}) => {
  const {
    colors,
    palette,
    radii,
    spacing
  } = core.useTheme();
  const list = adminUi_context_dist_keystone.useList(listKey);
  const [isCreateModalOpen, setIsCreateModalOpen] = React.useState(false);
  const router$1 = router.useRouter();
  return core.jsx("div", {
    css: {
      position: 'relative'
    }
  }, core.jsx(adminUi_router_dist_keystone.Link, {
    href: `/${list.path}`,
    css: {
      backgroundColor: colors.background,
      borderColor: colors.border,
      borderRadius: radii.medium,
      borderWidth: 1,
      // boxShadow: shadow.s100,
      display: 'inline-block',
      minWidth: 280,
      padding: spacing.large,
      textDecoration: 'none',
      ':hover': {
        borderColor: palette.blue400
      },
      ':hover h3': {
        textDecoration: 'underline'
      }
    }
  }, core.jsx("h3", {
    css: {
      margin: `0 0 ${spacing.small}px 0`
    }
  }, list.label, " "), count.type === 'success' ? core.jsx("span", {
    css: {
      color: colors.foreground,
      textDecoration: 'none'
    }
  }, count.count, " item", count.count !== 1 ? 's' : '') : count.type === 'error' ? count.message : count.type === 'loading' ? core.jsx(loading.LoadingDots, {
    label: `Loading count of ${list.plural}`,
    size: "small",
    tone: "passive"
  }) : 'No access'), core.jsx(CreateButton, {
    title: `Create ${list.singular}`,
    disabled: isCreateModalOpen,
    onClick: () => {
      setIsCreateModalOpen(true);
    }
  }, core.jsx(PlusIcon.PlusIcon, {
    size: "large"
  }), core.jsx(core.VisuallyHidden, null, "Create ", list.singular)), core.jsx(modals.DrawerController, {
    isOpen: isCreateModalOpen
  }, core.jsx(CreateItemDrawer.CreateItemDrawer, {
    listKey: list.key,
    onCreate: ({
      id
    }) => {
      router$1.push(`/${list.path}/${id}`);
    },
    onClose: () => {
      setIsCreateModalOpen(false);
    }
  })));
};

const CreateButton = props => {
  const theme = core.useTheme();
  return core.jsx("button", _extends({
    css: {
      alignItems: 'center',
      backgroundColor: theme.palette.neutral400,
      border: 0,
      borderRadius: theme.radii.xsmall,
      color: 'white',
      cursor: 'pointer',
      display: 'flex',
      height: 32,
      justifyContent: 'center',
      outline: 0,
      position: 'absolute',
      right: theme.spacing.large,
      top: theme.spacing.large,
      transition: 'background-color 80ms linear',
      width: 32,
      '&:hover, &:focus': {
        backgroundColor: theme.tones.positive.fill[0]
      }
    }
  }, props));
};

const HomePage = () => {
  const {
    adminMeta: {
      lists
    },
    visibleLists
  } = adminUi_context_dist_keystone.useKeystone();
  const query = React.useMemo(() => client.gql`
    query {
      keystone {
        adminMeta {
          lists {
            key
            fields {
              path
              createView {
                fieldMode
              }
            }
          }
        }
      }
      ${Object.entries(lists).map(([listKey, list]) => `${listKey}: ${list.gqlNames.listQueryCountName}`).join('\n')}
    }`, [lists]);
  let {
    data,
    error
  } = client.useQuery(query, {
    errorPolicy: 'all'
  });
  const dataGetter$1 = dataGetter.makeDataGetter(data, error === null || error === void 0 ? void 0 : error.graphQLErrors);
  return core.jsx(PageContainer.PageContainer, {
    header: core.jsx(core.Heading, {
      type: "h3"
    }, "Dashboard")
  }, visibleLists.state === 'loading' ? core.jsx(core.Center, {
    css: {
      height: `calc(100vh - ${PageContainer.HEADER_HEIGHT}px)`
    }
  }, core.jsx(loading.LoadingDots, {
    label: "Loading lists",
    size: "large",
    tone: "passive"
  })) : core.jsx(core.Inline, {
    as: "ul",
    gap: "large",
    paddingY: "xlarge",
    css: {
      paddingLeft: '0px',
      marginBottom: '0px'
    }
  }, (() => {
    if (visibleLists.state === 'error') {
      return core.jsx("span", {
        css: {
          color: 'red'
        }
      }, visibleLists.error instanceof Error ? visibleLists.error.message : visibleLists.error[0].message);
    }

    return Object.keys(lists).map(key => {
      var _result$errors;

      if (!visibleLists.lists.has(key)) {
        return null;
      }

      const result = dataGetter$1.get(key); // TODO: Checking based on the message is bad, but we need to revisit GraphQL errors in
      // Keystone to fix it and that's a whole other can of worms...

      if (((_result$errors = result.errors) === null || _result$errors === void 0 ? void 0 : _result$errors[0].message) === 'You do not have access to this resource') {
        return core.jsx(ListCard, {
          count: {
            type: 'no-access'
          },
          key: key,
          listKey: key
        });
      }

      return core.jsx(ListCard, {
        count: data ? result.errors ? {
          type: 'error',
          message: result.errors[0].message
        } : {
          type: 'success',
          count: data[key]
        } : {
          type: 'loading'
        },
        key: key,
        listKey: key
      });
    });
  })()));
};

exports.HomePage = HomePage;
