// @keystone-next/keystone/admin-ui/router/dist/keystone.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectWithoutProperties = require('@babel/runtime/helpers/objectWithoutProperties');
var router = require('next/router');
var Link$1 = require('next/link');
var React = require('react');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var Link__default = /*#__PURE__*/_interopDefault(Link$1);
var React__default = /*#__PURE__*/_interopDefault(React);

const _excluded = ["href", "as", "replace", "scroll", "shallow", "prefetch"];
const Link = _ref => {
  let {
    href,
    as,
    replace,
    scroll,
    shallow,
    prefetch
  } = _ref,
      props = _objectWithoutProperties(_ref, _excluded);

  return /*#__PURE__*/React__default['default'].createElement(Link__default['default'], {
    href: href,
    as: as,
    replace: replace,
    scroll: scroll,
    shallow: shallow,
    prefetch: prefetch
  }, /*#__PURE__*/React__default['default'].createElement("a", props));
};

Object.defineProperty(exports, 'Router', {
  enumerable: true,
  get: function () {
    return router.Router;
  }
});
Object.defineProperty(exports, 'createRouter', {
  enumerable: true,
  get: function () {
    return router.createRouter;
  }
});
Object.defineProperty(exports, 'makePublicRouterInstance', {
  enumerable: true,
  get: function () {
    return router.makePublicRouterInstance;
  }
});
Object.defineProperty(exports, 'useRouter', {
  enumerable: true,
  get: function () {
    return router.useRouter;
  }
});
Object.defineProperty(exports, 'withRouter', {
  enumerable: true,
  get: function () {
    return router.withRouter;
  }
});
exports.Link = Link;
