// @keystone-next/keystone/___internal-do-not-use-will-break-in-patch/admin-ui/next-config/dist/keystone.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _objectSpread = require('@babel/runtime/helpers/objectSpread2');
var Path = require('path');
var withPreconstruct = require('@preconstruct/next');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var Path__default = /*#__PURE__*/_interopDefault(Path);
var withPreconstruct__default = /*#__PURE__*/_interopDefault(withPreconstruct);

const config = withPreconstruct__default['default']({
  typescript: {
    ignoreBuildErrors: true
  },
  eslint: {
    ignoreDuringBuilds: true
  },

  webpack(config, {
    isServer
  }) {
    config.resolve.alias = _objectSpread(_objectSpread({}, config.resolve.alias), {}, {
      react: Path__default['default'].dirname(require.resolve('react/package.json')),
      'react-dom': Path__default['default'].dirname(require.resolve('react-dom/package.json')),
      '@keystone-next/keystone': Path__default['default'].dirname(require.resolve('@keystone-next/keystone/package.json'))
    });

    if (isServer) {
      config.externals = [...config.externals, /@keystone-next\/keystone(?!\/___internal-do-not-use-will-break-in-patch\/admin-ui\/id-field-view|\/fields\/types\/[^\/]+\/views)/, '.prisma/client'];
    }

    return config;
  }

});

exports.config = config;
